### PR TITLE
FCS: add event when FSI emits

### PR DESCRIPTION
**Pull request by goswinr**
On bigger fsx scripts, the compilation can take a few seconds. This new event can be used by any UI hosting FCS to indicate when the actual fsi evaluation starts.
I'm currently using `unit` as Event args. I think this might be good enough. 
I hope you can accept this pull request. 
(I am using FSI with my own WPF scripting editor. I want to publish it in on GitHub soon.)

### CODE REVIEW

File: src/Compiler/Interactive/fsi.fsi [hunk 1]
```diff
@@ -92,6 +92,10 @@ type public FsiEvaluationSessionHostConfig =
     /// Hook for listening for evaluation bindings
     member OnEvaluation: IEvent<EvaluationEventArgs>
 
+    /// Hook for listening for start of emitting dynamic Assemblies. After Compilation has finished.
+    /// Can be used by an UI hosting the compiler service to indicate that fsi compilation is finished and fsi evaluation is starting.
+    member OnEmit: IEvent<unit>
```
Comment (psfinaki) [inline src/Compiler/Interactive/fsi.fsi:97]: How about `OnEmitStart`?

File: src/Compiler/Interactive/fsi.fs [hunk 2]
```diff
@@ -582,9 +584,15 @@ type FsiEvaluationSessionHostConfig() =
     /// Hook for listening for evaluation bindings
     member _.OnEvaluation = evaluationEvent.Publish
 
+    /// Hook for listening for start of emitting dynamic Assemblies. After Compilation has finished.
```
Comment (psfinaki) [inline src/Compiler/Interactive/fsi.fs:587]:
```suggestion
    /// Hook for listening for start of emitting dynamic assemblies, after compilation has finished.
```

**Review by psfinaki — COMMENTED** [review]
This makes sense to me. Do you have a guesstimate on how big the script should be to make this distinct event useful?

**Comment by smoothdeveloper** [comment]
> Do you have a guesstimate on how big the script should be to make this distinct event useful?

I think the right approach would be to notify the user if it takes more than 0.5 seconds and start showing the status.

We should look at how the jupyter notebook python extension in VSCode shows the status of evaluation of cells as guidance in how/when to show the status of the current evaluation.

image

image

But in context of current FSI in VS, I assume it would be displayed in the status bar.

@goswinr congrats for working on FSI wrapper and thanks for contributing to FCS :)

**Comment by goswinr** [comment]
@psfinaki I have a few scripts about 2000 lines long. 
I use them interactively in Rhino3D. These scripts often start with some UI dialog. There is a 2-4 second wait for the dialog to show up. 
Having this event, I can distinguish between some code still compiling or some evaluation taking rather long. 

@smoothdeveloper I don't think this PR should do any timing! It should just always trigger an event. Just like the `OnEvaluation` Event.
It should be up for the users of fsi (my WPF app, Visual Studio, jupyter) to do timing and react in the UI. 

Thanks for your congratulations, I have been doing this since the Tsunami editor https://www.youtube.com/watch?v=ZY-bvZZZZnE discontinued. 
So glad I am not scripting in Python anymore !! I am too clumsy for that! I need a type checker !

**Comment by smoothdeveloper** [comment]
>  I don't think this PR should do any timing!

I was considering the implication of using the event you are adding in VS, as I think it would be a potential client for the new API, so I agree, the event added here don't do any timing.

If I understand correctly, there is already an event for the evaluation, and you are adding the one for only the "parse/optimize/emit IL" part, so I guess my remark was more for the overall lifecycle in terms of feedback to the user.

Maybe it would be better to have a single event to subscribe to, that sends the different status as it progresses:

```fsharp
type FSharpSubmissionLifeCycleStage =
| EmitStart
| EvaluationStart
| EvaluationCompleted of Result<...,...>
```

But I think it is outside the scope of this PR, and would require more ponderation on the design of FCS clients integrating FSI workflow.

**Comment by T-Gro** [comment]
Maybe you didn't know this:

Existing `fsi.fs` code is already covered with various calls to `ReportTime`, which underneath is an `System.Diagnostics.Activity` trace which can be listened to.

You can create a `System.Diagnostics.ActivityListener`, configure it to listen to `ActivityNames.ProfiledSourceName` string and you can be getting events from the processing alltogether.

**Comment by goswinr** [comment]
@T-Gro I did not know about ActivityListener. Thanks, this would work. But events are somtimes  triggered **twice** if there is a `#r "nuget: SomeLib"`
I guess the same would happen with the event in my PR ?
Would there be another possibility to have **a single event** after parse/optimize/emit and before evaluation?

I am using
```fsharp
Diagnostics.ActivitySource.AddActivityListener(
    new Diagnostics.ActivityListener(
        ShouldListenTo  = (fun a -> a.Name = ActivityNames.ProfiledSourceName), 
        Sample          = (fun _ -> Diagnostics.ActivitySamplingResult.AllData),
        ActivityStopped = (fun a -> printfn "ActivityStopped: %s" a.OperationName),
        ActivityStarted = (fun a -> printfn "ActivityStarted: %s" a.OperationName))
    )
```

for 
```
#r "nuget: FsEx"
let x = 2
```
I get :
```
ActivityStarted: Optimizations
ActivityStopped: Optimizations
ActivityStarted: Ending Optimizations
ActivityStopped: Ending Optimizations
(...) 
ActivityStarted: PDB: Defined 1 documents
ActivityStopped: PDB: Defined 1 documents
(...) 
ActivityStopped: Reflection.Emit
ActivityStarted: Run Bindings
ActivityStopped: Run Bindings
ActivityStarted: Optimizations
ActivityStopped: Optimizations
ActivityStarted: Ending Optimizations
ActivityStopped: Ending Optimizations
(...) 
ActivityStarted: PDB: Sorted 2 methods
ActivityStopped: PDB: Sorted 2 methods
(...) 
ActivityStopped: Reflection.Emit
ActivityStarted: Run Bindings
```

**Comment by T-Gro** [comment]
> @T-Gro I did not know about ActivityListener. Thanks, this would work. But events are somtimes triggered **twice** if there is a `#r "nuget: SomeLib"` I guess the same would happen with the event in my PR ? Would there be another possibility to have **a single event** after parse/optimize/emit and before evaluation?
> 
> I am using
> 
> ```fsharp
> Diagnostics.ActivitySource.AddActivityListener(
>     new Diagnostics.ActivityListener(
>         ShouldListenTo  = (fun a -> a.Name = ActivityNames.ProfiledSourceName), 
>         Sample          = (fun _ -> Diagnostics.ActivitySamplingResult.AllData),
>         ActivityStopped = (fun a -> printfn "ActivityStopped: %s" a.OperationName),
>         ActivityStarted = (fun a -> printfn "ActivityStarted: %s" a.OperationName))
>     )
> ```
> 
> for
> 
> ```
> #r "nuget: FsEx"
> let x = 2
> ```
> 
> I get :
> 
> ```
> ActivityStarted: Optimizations
> ActivityStopped: Optimizations
> ActivityStarted: Ending Optimizations
> ActivityStopped: Ending Optimizations
> (...) 
> ActivityStarted: PDB: Defined 1 documents
> ActivityStopped: PDB: Defined 1 documents
> (...) 
> ActivityStopped: Reflection.Emit
> ActivityStarted: Run Bindings
> ActivityStopped: Run Bindings
> ActivityStarted: Optimizations
> ActivityStopped: Optimizations
> ActivityStarted: Ending Optimizations
> ActivityStopped: Ending Optimizations
> (...) 
> ActivityStarted: PDB: Sorted 2 methods
> ActivityStopped: PDB: Sorted 2 methods
> (...) 
> ActivityStopped: Reflection.Emit
> ActivityStarted: Run Bindings
> ```

Might be for desing-time scenarios.
The Activities also have parent-child relationships - if  there are two passes of the code, I think we would be able to report them individually, and then each time reporting would belong to a different parent.

But personally I do not know where/how the two passes originate, but it should for sure be doable.

Alternatively, if the current granularity of the activities is not sufficient, we could add additional ones to make it a batter fit for FSI.
The good thing is, unless being listened to, Activities are basically a zero-cost abstraction (and even when being listened to, they are really lightweight).

**Comment by T-Gro** [comment]
@goswinr: I think it should be possible to emit a "parent" activity right at the place where you added the `fsi.TriggerEmitStart()` line, and then stop it after emit ends.

That way there would be a separating activity between the two groups of telemetry, and the `.Parent` property of an activity should tell which is which.

Would this approach work for you?

**Comment by goswinr** [comment]
@T-Gro yes, that could work. So the actual emitting event that I am after would have no parent?

**Comment by vzarytovskii** [comment]
Should this be re-done/updated to use activities?

**Comment by goswinr** [comment]
I think I can work with the build in Activity Tracker. So this can be closed. Thank you for your support!

**Pull request closed without merging**